### PR TITLE
NEW add members types ldap group management

### PR DESCRIPTION
**Pull request by hregis**
NEW add members types ldap group management
NEW big refactorization and uniformize code of member type
NEW Ping ldap server before ldap_connect for avoid waiting
NEW add member type triggers

@eldy

### CODE REVIEW

File: htdocs/adherents/class/adherent.class.php [hunk 1]
```diff
@@ -714,109 +714,108 @@ function update_end_date($user)
...
-        	dol_syslog(get_class($this)."::delete", LOG_DEBUG);
-        	$resql=$this->db->query($sql);
-        	if (! $resql)
-        	{
-        		$error++;
-        		$this->error .= $this->db->lasterror();
-        		$errorflag=-5;
-        	}
-        }
+		if (! $error && ! $notrigger)
```
Comment (eldy) [inline htdocs/adherents/class/adherent.class.php]: For the "delete", the trigger must be called before the delete of object (can be after child tables) but must be before main table (this is the opposite for create and update).

Comment (hregis) [reply htdocs/adherents/class/adherent.class.php]: @eldy 
if the trigger is before, the member is always considered to be present in the member type !

**Comment by hregis** [comment]
@eldy 
ok I have done otherwise

**Comment by hregis** [comment]
@eldy 
can you accept this commit that I can advance on ldap? thanks

**Pull request merged**